src/main/java/uk/gov/register/core/TransactionalMemoizationStore.java

```diff
@@ -0,0 +1,49 @@
+package uk.gov.register.core;
+
+import uk.gov.verifiablelog.store.memoization.MemoizationStore;
+import java.util.HashMap;
+
+public class TransactionalMemoizationStore implements MemoizationStore {
+    private final MemoizationStore memoizationStore;
+    private final HashMap<Integer, HashMap<Integer, byte[]>> stagedHashes;
+
+    public TransactionalMemoizationStore(MemoizationStore memoizationStore) {
+        this.memoizationStore = memoizationStore;
+        this.stagedHashes = new HashMap<>();
+    }
+
+    @Override
+    public void put(Integer start, Integer size, byte[] value) {
+        if (!stagedHashes.containsKey(start)) {
+            stagedHashes.put(start, new HashMap<>());
+        }
+
+        HashMap<Integer, byte[]> hashesBySize = stagedHashes.get(start);
+        hashesBySize.put(size, value);
+    }
+
+    @Override
+    public byte[] get(Integer start, Integer size) {
+        byte[] hash = null;
+
+        if (stagedHashes.containsKey(start)) {
+            hash = stagedHashes.get(start).get(size);
```

Comment:
in which case, I'd find `if (hash != null) { return hash; }` less confusing here than the ternary currently used below